libshaderc/include/shaderc.h

```diff
@@ -141,6 +141,55 @@ void shaderc_compile_options_set_disassembly_mode(
 void shaderc_compile_options_set_forced_version_profile(
     shaderc_compile_options_t options, int version, shaderc_profile profile);
 
+// When including a file, libshaderc initiates a query to get the full
+// path and the content of a file. The query is initiated through calling a
+// function pointer, which is a client-side implemented reponse method. The
+// response method should follow the specified function signature below, and
+// it should be passed to libshaderc through corresponding setter functions.
+// When the including of a file is done, libshaderc will call a client-side
+// implemented method to clean up the resources used for the including
+// process.  Client should implement the clean up method and pass it as a
+// callback function pointer to libshaderc along with the response method.
+
+// The struct that contains the information to be returned to the includer.
+// The client-side implemented response method should return this struct. And
+// the underlying data is owned by client code.
+struct shaderc_includer_fullpath_content {
+  const char* fullpath;
+  const size_t fullpath_length;
+  const char* content;
+  const size_t content_length;
+};
+
+// The function signature of the client-side implemented reponse method. It
+// returns shaderc_includer_fullpath_content struct.
+typedef shaderc_includer_fullpath_content (
+    *shaderc_includer_fullpath_and_content_getter_fn)(void* user_data,
+                                                      const char* filename);
+
+// The function signature of the client-side implemented clean-up method.
+// Includer will call this callback function when the including process is done
+// with the fullpath and content data.
+typedef void (*shaderc_includer_fullpath_and_content_finalizer_fn)(
```

Comment:
This seems overly complex.  Why do you need to provide the original filename?  

I expected the original query to return a *pointer* to a shadec_includer_response, then this "finalizer" would really be
  typedef void (*shaderc_includer_response_release_fn)(void* user_data, shaderc_includer_response* data);